### PR TITLE
Basic implementation of hot-loading/mods

### PR DESCRIPTION
**Pull request by martinrlilja**
Addresses parts of #93.

## What I've done:

 * Implemented package format. Skipped dependencies.
 * Loading of packages from disk.
 * Loading of dylibs, registered from packages.
 * Interfaces for mods to implement in order to be loaded.
 * A tiny proof of concept mod.

## How to use:

 1. Run the game with `cargo run`.
 2. If everything went well, you should see the text `my mod was loaded!` in the terminal.

## What's missing:

 * Only tested on Linux x86_64, and MacOS.
 * ~~Loading of whole directories of mods.~~
 * ~~Dependencies~~, and loads of other things.

This is just a first prototype though, but I wanted to put it up to get some feedback before I continue.

### CODE REVIEW

**Comment by Herbstein** [comment]
I'll test this on a Windows 10 64-bit laptop in the afternoon. Impressive work!

**Comment by aeickhoff** [comment]
Great work!

FYI: on Mac, everything compiles, but I get this error on startup, which I don't really understand:

```
dyld: Library not loaded: @rpath/librustc_metadata-b03b26491b91b850.dylib
  Referenced from: /Users/anselm/Dropbox (Personal)/Data/Projekte/cb/rust/target/release/citybound
  Reason: image not found
```

I can't find this `librustc_metadata-b03b26491b91b850.dylib` in `target/release/deps` and don't know where else it could be.

**Comment by martinrlilja** [comment]
@aeickhoff That's weird, could you try to remove `extern crate rustc_metadata;` from `weaver/lib.rs`?

File: lib/weaver/src/weaver.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+        let mut path = path.clone();
+        path.push(package.name());
+        path.set_extension(EXT);
+
+        if !path.exists() {
+            return Err(format!("module does not exist {:?}", path));
+        }
+        if !path.is_file() {
+            return Err(format!("module does not a file {:?}", path));
```
Comment (chances) [inline lib/weaver/src/weaver.rs]: Should be, "module _is_ not a file ..."?

Comment (martinrlilja) [inline lib/weaver/src/weaver.rs]: Yup, fixed :smile:

**Comment by aeickhoff** [comment]
> Out of curiosity, what happens if you just import extern crate rustc_driver; and call rustc_driver::main(); somewhere and run that?

Tried that, got the same error as above.

**Comment by martinrlilja** [comment]
@aeickhoff Okay, I'm fairly certain that the error is because the library is not in your path. You should be able to verify it with `otool` on mac. So, try my latest version, run it with `cargo run`, it should correctly set the path to your rustup install. Otherwise you might have to set `DYLD_LIBRARY_PATH`.

**Comment by aeickhoff** [comment]
By setting `DYLD_LIBRARY_PATH` to the lib folder in my dustup install, it works indeed! (but crashes when quitting the application)

also, FYI: Mac doesn't recognise the `-enable-new-dtags` ld option

**Comment by martinrlilja** [comment]
That's great! I think I will make it so that the ld option only runs on linux.

The crash on exit is a bit worrying, it might be either because of this: https://github.com/citybound/citybound/pull/95/files#diff-98e384c6f9803e1156e01ae31bb92c57R163 or because of this: https://github.com/nagisa/rust_libloading/issues/5

**Comment by martinrlilja** [comment]
@chances Just to make it more obvious that we are talking about the root of everything. Kind of like URLs. We could also make it so that a single forward slash refers to the root of the current source. It's definitely something we should discuss more about to find something that works the best.

**Comment by aeplay** [comment]
Archiving this for now, I'll include you guys in discussion regarding the new mod-from-inside-the-game idea when it's time!

**Pull request closed without merging**